stream/src/public_key/mod.rs

```diff
@@ -1,67 +1,98 @@
-//!
-//! In any handshake, the dialer is the party that attempts to connect to some known address/identity (public key)
-//! and the recipient of this connection is the listener. Upon forming a TCP connection, the dialer sends a signed
-//! handshake message to the listener. Besides the signature and the public key of the dialer, the handshake message
-//! contains:
-//!
-//! - The receiver's public key.
-//! - The sender's ephemeral public key (used to establish a shared secret).
-//! - The current timestamp (used to prevent replay attacks).
-//!
-//! The listener verifies the public keys are well-formatted, the timestamp is valid (not too old/not too far in the future),
-//! and that the signature is valid. If all these checks pass, the listener checks to see if it is already connected or dialing
-//! this peer. If it is, it drops the connection. If it isn't, it sends back its own signed handshake message (same as above)
-//! and considers the connection established.
-//!
-//! Upon receiving the listener's handshake message, the dialer verifies the same data as the listener and additionally verifies
-//! that the public key returned matches what they expected at the address. If all these checks pass, the dialer considers the
-//! connection established. If not, the dialer drops the connection.
-//!
-//! To better protect against malicious peers that create and/or accept connections but do not participate in handshakes,
-//! a configurable deadline is enforced for any handshake to be completed. This allows for the underlying runtime to maintain
-//! a standard read/write timeout for connections without making it easier for malicious peers to keep useless connections open.
+//! A 3-message handshake is performed to mutually authenticate and establish a shared secret.
+//! This handshake is used instead of TLS, Noise, WireGuard, etc. because:
+//! - It supports arbitrary cryptographic schemes
+//! - No protocol negotiation (only one way to connect)
+//! - Simple implementation (few hundred lines of code)
+//! - Can be simulated deterministically
+//!
+//! The **dialer** initiates connection to a known address/identity (public key)
+//! and the **listener** receives this connection. The protocol works as follows:
+//!
+//! ### Message 1: Dialer → Listener (Initial Handshake)
+//! Upon establishing a TCP connection, the dialer sends a signed handshake message to the listener.
+//! This message contains the dialer's signature and static public key, plus:
+//!
+//! - The listener's expected public key
+//! - The dialer's ephemeral public key (for establishing a shared secret)
+//! - The current timestamp (for replay attack prevention)
+//!
+//! ### Message 2: Listener → Dialer (Response with Key Confirmation)
+//! The listener verifies:
+//! - Public keys are well-formatted
+//! - Timestamp is valid (not too old, not too far in the future)
+//! - Signature is valid
+//! - It is not already connected to or dialing this peer
```

Comment:
This is not the concern of `stream`. I think we could put this in italics but putting it here could lead to someone using `stream` not preventing this.